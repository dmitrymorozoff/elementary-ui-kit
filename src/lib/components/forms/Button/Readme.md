### Basic Usage:

```jsx
	<Button>Default</Button>
	<Button type="primary">Primary</Button>
	<Button type="success">Success</Button>
	<Button type="warning">Warning</Button>
	<Button type="danger">Danger</Button>
```

### Round Buttons:

```jsx
	<Button round>Default</Button>
	<Button type="primary" round>
		Primary
	</Button>
	<Button type="success" round>
		Success
	</Button>
	<Button type="warning" round>
		Warning
	</Button>
	<Button type="danger" round>
		Danger
	</Button>
```

### Disabled Buttons:

The `disabled` attribute determines if the button is disabled.

```jsx
	<Button disabled={true}>Default</Button>
	<Button disabled={true} type="primary">
		Primary
	</Button>
	<Button disabled={true} type="success">
		Success
	</Button>
	<Button disabled={true} type="warning">
		Warning
	</Button>
	<Button disabled={true} type="danger">
		Danger
	</Button>
```

### Text Buttons:

Buttons without border and background.

```jsx
	<Button type="text">Default</Button>
	<Button type="text" disabled={true}>
		Primary
	</Button>
```

### Sizes:

Besides default size, Button component provides three additional sizes.

```jsx
	<Button>Default</Button>
	<Button size="medium" type="primary">Medium</Button>
	<Button size="small" type="primary">Small</Button>
	<Button size="mini" type="primary">Mini</Button>
	<br/>
	<Button round>Default</Button>
	<Button size="medium" round type="danger">
		Medium
	</Button>
	<Button size="small" round type="danger">
		Small
	</Button>
	<Button size="mini" round type="danger">
		Mini
	</Button>
```
