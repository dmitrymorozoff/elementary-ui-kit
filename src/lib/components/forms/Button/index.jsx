import React from "react";
import styled from "styled-components";
import { lighten, darken } from "polished";
import { PropTypes } from "prop-types";

const defaultColors = props => {
    if (props.type === "primary") return props.theme.type.primary;
    if (props.type === "success") return props.theme.type.success;
    if (props.type === "warning") return props.theme.type.warning;
    if (props.type === "danger") return props.theme.type.danger;
    return "transparent";
};

const ButtonElement = styled.button`
    font-family: "Lato", sans-serif;
    font-size: 15px;
    border: 2px solid ${props => props.theme.type.default};
    background-color: transparent;
    line-height: 1;
    white-space: nowrap;
    box-sizing: border-box;
    outline: none;
    margin: ${props => props.margin};
    width: ${props => (props.fullWidth ? "100%" : "auto")};
    text-align: center;
    padding: ${props => {
        if (props.size === "medium") return props.theme.size.medium;
        if (props.size === "small") return props.theme.size.small;
        if (props.size === "mini") return props.theme.size.mini;
        return props.theme.size.default;
    }};
    transition: 0.15s;
    color: "#333";
    cursor: pointer;
    border: 1px solid
        ${props => {
            if (props.type === "primary") return props.theme.type.primary;
            if (props.type === "success") return props.theme.type.success;
            if (props.type === "warning") return props.theme.type.warning;
            if (props.type === "danger") return props.theme.type.danger;
            if (props.type === "text") return props.theme.type.text;
            return props.theme.type.default;
        }};
    color: ${props => {
        if (
            props.type === "primary" ||
            props.type === "success" ||
            props.type === "warning" ||
            props.type === "danger"
        ) {
            return "#fff";
        }
        if (props.type === "text") {
            return props.theme.type.primary;
        }
    }};
    background-color: ${defaultColors};
    border-radius: ${props => {
        if (props.round) {
            return "25px";
        }
    }};
    box-shadow: 0 8px 25px -6px ${defaultColors};
    &:hover {
        background-color: ${props => {
            if (props.type === "primary") return props.theme.type.primary && lighten(0.05, props.theme.type.primary);
            if (props.type === "success") return props.theme.type.success && lighten(0.05, props.theme.type.success);
            if (props.type === "warning") return props.theme.type.warning && lighten(0.05, props.theme.type.warning);
            if (props.type === "danger") return props.theme.type.danger && lighten(0.05, props.theme.type.danger);
            if (props.type === "text") return props.theme.type.text;
            return props.theme.type.primary && lighten(0.35, props.theme.type.primary);
        }};
        color: ${props => {
            if (!props.type) {
                return props.theme.type.primary;
            }
        }};
        border: 1px solid
            ${props => {
                if (!props.type) {
                    return props.theme.type.primary;
                }
                return "transparent";
            }};
    }
    &:active {
        background-color: ${props => {
            if (props.type === "primary") return props.theme.type.primary && darken(0.05, props.theme.type.primary);
            if (props.type === "success") return props.theme.type.success && darken(0.05, props.theme.type.success);
            if (props.type === "warning") return props.theme.type.warning && darken(0.05, props.theme.type.warning);
            if (props.type === "danger") return props.theme.type.danger && darken(0.05, props.theme.type.danger);
            if (props.type === "text") return props.theme.type.text;
            return props.theme.type.primary && lighten(0.32, props.theme.type.primary);
        }};
        color: ${props => {
            if (props.type === "text") {
                return props.theme.type.primary && darken(0.15, props.theme.type.primary);
            }
        }};
        box-shadow: 0 4px 15px -6px ${defaultColors};
    }
    &:focus {
        border-style: dashed;
    }
    &:disabled {
        cursor: not-allowed;
        opacity: 0.65;
        &:hover {
            background-color: ${defaultColors};
        }
        color: ${props => {
            if (props.type === "text") {
                return props.theme.type.default;
            }
        }};
    }
`;

export class Button extends React.Component {
    render() {
        return <ButtonElement {...this.props} />;
    }
}

Button.defaultProps = {
    theme: {
        type: {
            default: "#d8dadd",
            primary: "#3A77F8",
            success: "#4FD496",
            warning: "#FAAF27",
            danger: "#FE0C0D",
            text: "transparent",
        },
        size: {
            default: "15px 36px",
            medium: "12px 20px",
            small: "8px 18px",
            mini: "6px 14px",
        },
    },
    leftIconComponent: null,
    rightIconComponent: null,
    fullWidth: false,
    margin: 0,
};

Button.propTypes = {
    leftIconComponent: PropTypes.node,
    rightIconComponent: PropTypes.node,
    theme: PropTypes.object,
    children: PropTypes.node,
    disabled: PropTypes.bool,
    fullWidth: PropTypes.bool,
    round: PropTypes.bool,
    type: PropTypes.oneOf(["default", "primary", "success", "warning", "danger", "text"]),
    size: PropTypes.oneOf(["medium", "small", "mini"]),
    margin: PropTypes.string,
};
