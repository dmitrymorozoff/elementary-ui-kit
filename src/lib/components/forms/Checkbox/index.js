import React from "react";
import styled from "styled-components";
import { PropTypes } from "prop-types";

const getPropsColor = props => props.color;

const CheckboxWrapper = styled.label`
    font-family: "Roboto", sans-serif;
    position: relative;
    display: inline-block;
    font-size: 15px;
    height: 25px;
    padding-left: ${props => (props.label ? "32px" : "25px")};
    line-height: 25px;
    opacity: ${props => (props.disabled ? "0.65" : "1")};
    cursor: pointer;
`;

const CheckboxElement = styled.input.attrs({
    type: "checkbox",
})`
    transition: 0.2s;
    display: none;
`;

const Box = styled.span`
    position: absolute;
    width: 25px;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    transition: 0.4s;
    cursor: pointer;
    background-color: #d8dadd;
    border-radius: 30px;
    ${CheckboxElement}:checked + & {
        background-color: ${getPropsColor};
        box-shadow: 0 4px 25px -6px ${getPropsColor};
    }
    ${CheckboxElement}:focus + & {
        box-shadow: 0 0 1px ${getPropsColor};
    }
    &:before {
        position: absolute;
        content: "";
        left: 4px;
        bottom: 4px;
        transition: 0.4s;
        height: 17px;
        width: 17px;
        background-color: white;
        border-radius: 50%;
        ${CheckboxElement}:hover + & {
            transform: scale(0.75);
        }
        ${CheckboxElement}:checked + & {
            transform: scale(1);
        }
    }
`;

export class Checkbox extends React.Component {
    render() {
        const { disabled, color, label, className } = this.props;
        return (
            <CheckboxWrapper
                disabled={disabled}
                label={label}
                className={className}
            >
                <CheckboxElement {...this.props} />
                <Box color={color} />
                {label}
            </CheckboxWrapper>
        );
    }
}

Checkbox.defaultProps = {
    onChange: () => {},
    color: "#3a77f8",
    label: "",
};

Checkbox.propTypes = {
    checked: PropTypes.bool,
    defaultChecked: PropTypes.bool,
    onChange: PropTypes.func,
    name: PropTypes.string,
    color: PropTypes.string,
    disabled: PropTypes.bool,
    label: PropTypes.string,
    className: PropTypes.string,
};
