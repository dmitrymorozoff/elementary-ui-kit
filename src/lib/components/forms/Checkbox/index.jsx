import React from "react";
import styled from "styled-components";
import { PropTypes } from "prop-types";

const CheckboxWrapper = styled.label`
    position: relative;
    display: inline-block;
    width: 50px;
    height: 25px;
    margin: 8px;
`;

const CheckboxElement = styled.input.attrs({
    type: "checkbox",
})`
    display: none;
`;

const Slider = styled.span`
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    transition: 0.4s;
    cursor: pointer;
    background-color: #d8dadd;
    border-radius: 30px;
    ${CheckboxElement}:checked + & {
        background-color: #3a77f8;
        box-shadow: 0 4px 25px -6px #3a77f8;
    }
    ${CheckboxElement}:focus + & {
        box-shadow: 0 0 1px #3a77f8;
    }
    &:before {
        position: absolute;
        content: "";
        left: 4px;
        bottom: 4px;
        transition: 0.4s;
        height: 17px;
        width: 17px;
        background-color: white;
        border-radius: 50%;
        ${CheckboxElement}:checked + & {
            transform: translateX(25px);
        }
    }
`;

export class Checkbox extends React.Component {
    render() {
        return (
            <CheckboxWrapper>
                <CheckboxElement {...this.props} />
                <Slider />
            </CheckboxWrapper>
        );
    }
}

Checkbox.defaultProps = {};

Checkbox.propTypes = {
    checked: PropTypes.bool,
};
