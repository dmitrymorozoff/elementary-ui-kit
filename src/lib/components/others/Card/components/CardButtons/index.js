import styled from "styled-components";
import { PropTypes } from "prop-types";

/** @component */
export const CardButtons = styled.div`
    display: flex;
    justify-content: space-between;
    margin-top: auto;
    flex-wrap: wrap;
`;

CardButtons.propTypes = {
    className: PropTypes.string,
};
