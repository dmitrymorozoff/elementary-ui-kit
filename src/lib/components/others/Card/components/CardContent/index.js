import styled from "styled-components";
import { PropTypes } from "prop-types";

/** @component */
export const CardContent = styled.div`
    font-family: "Roboto", sans-serif;
    line-height: 1.5;
    display: flex;
    flex-direction: column;
    width: 100%;
    height: 100%;
    padding: ${props => props.padding};
    box-sizing: border-box;
`;

CardContent.defaultProps = {
    padding: "15px 5px",
};

CardContent.propTypes = {
    padding: PropTypes.string,
    className: PropTypes.string,
};
