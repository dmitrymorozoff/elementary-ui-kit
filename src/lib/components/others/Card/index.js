import React from "react";
import styled from "styled-components";
import { PropTypes } from "prop-types";

export const Card = styled.div`
    display: flex;
    flex-direction: column;
    align-items: center;
    min-width: 275px;
    min-height: 260px;
    width: ${props => props.width};
    height: ${props => props.height};
    box-shadow: 0 8px 25px -6px #f2f2f6;
    position: relative;
    margin: ${props => (props.withImage ? "45px 10px 0 10px" : "10px")};
    border: 1px solid #f2f2f5;
    box-sizing: border-box;
    box-shadow: 0 15px 20px -15px rgba(0, 0, 0, 0.15), 0 55px 50px -35px rgba(0, 0, 0, 0.15);
    padding: 15px;
    transition: 0.5s;
    cursor: ${props => (props.hover ? "pointer" : "default")};
    &:hover {
        box-shadow: ${props =>
            props.hover
                ? "0 15px 10px -15px rgba(0, 0, 0, 0.15), 0 35px 50px -35px rgba(0, 0, 0, 0.15)"
                : "0 15px 20px -15px rgba(0, 0, 0, 0.15), 0 55px 50px -35px rgba(0, 0, 0, 0.15)"};
    }
`;

Card.defaultProps = {
    width: "275px",
    height: "260px",
};

Card.propTypes = {
    children: PropTypes.node,
    width: PropTypes.string,
    height: PropTypes.string,
    hover: PropTypes.bool,
    withImage: PropTypes.bool,
};
