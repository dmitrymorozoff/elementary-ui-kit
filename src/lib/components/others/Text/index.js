import React from "react";
import styled from "styled-components";
import { PropTypes } from "prop-types";

export const Text = styled.p`
    font-family: "Lato", sans-serif;
    line-height: 1.5;
    font-size: ${props => props.fontSize};
    margin: 0;
    margin-bottom: ${props => (props.gutterBottom ? "8px" : 0)};
`;

Text.defaultProps = {
    fontSize: "14px",
    gutterBottom: false,
};

Text.propTypes = {
    fontSize: PropTypes.string,
    gutterBottom: PropTypes.bool,
};
