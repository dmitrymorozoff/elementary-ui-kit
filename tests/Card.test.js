import React from "react";
import { Card } from "../src/lib/components/others/Card";
import { Text } from "../src/lib/components/others/Text";
import { Button } from "../src/lib/components/forms/Button";
import { CardContent } from "../src/lib/components/others/Card/components/CardContent";
import { CardButtons } from "../src/lib/components/others/Card/components/CardButtons";

describe("<Card/>", () => {
    it("should render a default card", () => {
        const component = shallow(
            <Card>
                <CardContent>
                    <Text fontSize="28px" gutterBottom={true}>
                        Penguin
                    </Text>
                    <Text>
                        They live almost exclusively in the Southern Hemisphere, with only one species, the Galapagos
                        penguin, found north of the equator.
                    </Text>
                    <CardButtons>
                        <Button fullWidth={true}>Learn More</Button>
                    </CardButtons>
                </CardContent>
            </Card>,
        );
        expect(component).toMatchSnapshot();
    });
});
