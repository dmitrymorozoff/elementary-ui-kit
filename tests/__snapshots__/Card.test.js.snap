// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card/> should render a default card 1`] = `
<div
  className="sc-bdVaJa Wkyni"
  height="260px"
  width="275px"
>
  <styled.div
    padding="15px 5px"
  >
    <styled.p
      fontSize="28px"
      gutterBottom={true}
    >
      Penguin
    </styled.p>
    <styled.p
      fontSize="14px"
      gutterBottom={false}
    >
      They live almost exclusively in the Southern Hemisphere, with only one species, the Galapagos penguin, found north of the equator.
    </styled.p>
    <styled.div>
      <Button
        fullWidth={true}
        leftIconComponent={null}
        margin="0px"
        rightIconComponent={null}
        theme={
          Object {
            "size": Object {
              "default": "15px 36px",
              "medium": "12px 20px",
              "mini": "6px 14px",
              "small": "8px 18px",
            },
            "type": Object {
              "danger": "#FE0C0D",
              "default": "#d8dadd",
              "primary": "#3A77F8",
              "success": "#4FD496",
              "text": "transparent",
              "warning": "#FAAF27",
            },
          }
        }
      >
        Learn More
      </Button>
    </styled.div>
  </styled.div>
</div>
`;
